content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -579,41 +657,919 @@ sending values over the wire, the code should query the size of the
+in the same order as in the physical layout is convenient because you 
+can copy-and-paste the components in that second step, as a start. 
+
+Let's reuse that record type from the earlier discussion:
+
+.. code-block:: ada
+
+   type My_Int is range 1 .. 10;
+   
+   subtype S is Integer range 1 .. 10;
+   
+   type R is record
+      M : My_Int;
+      X : S;
+      B : Boolean;
+      C : Character;
+   end record;
+   
+The resulting layout is like so, assuming the compiler doesn't reorder 
+the components: 
+
+.. image:: images/unoptimized-record-component-order.png
+  :width: 600
+  :alt: Memory allocated to a record with unoptimized layout
+   
+As a result, R'Size will be 80 bits (10 bytes), but those last two bytes 
+will be will be allocated to objects, for an Object_Size of 96 bits (12 
+bytes). We'll change that with an explicit layout specification.
+
+Having declared the record type, the second step consists of defining 
+the corresponding record representation clause giving the components' 
+layout. The clause uses syntax that somewhat mirrors that of a record 
+type declaration. The components' names appear, as in a record type 
+declaration. But now, we don't repeat the components' types, instead we 
+give their relative positions within the record, in terms of a relative 
+offset that starts at zero. We also specify the bits we want them to 
+occupy within the storage elements starting at that offset. 
+
+.. code-block:: ada
+
+   for R use record
+      X at 0 range 0 .. 31;  -- note the order swap,
+      M at 4 range 0 ..  7;  -- with this component
+      B at 5 range 0 ..  7;
+      C at 6 range 0 ..  7;
+   end record;
+
+Now we'll get the optimized order, and we'll always get that order, or 
+the layout specification won't compile in the first place. In the 
+following diagram, both layouts, the default, and the one resulting from 
+the record representation clause, are depicted for comparison: 
+
+.. image:: images/record_layout_with_rep_clause.png
+  :width: 600
+  :alt: Memory allocated to a record with optimized layout
+  
+R'Size will be 56 bits (7 bytes), but that last padding byte will also 
+be allocated to objects, so the Object_Size will be 64 bits (8 bytes). 
+  
+Notice how we gave each component an offset, after the reserved word 
+"at". These offsets are in terms of storage elements, and specify their 
+positions within the record object as a whole. They are relative to the 
+beginning of the memory allocated to the record object so they are 
+numbered starting at zero. We want the X component to be the very first 
+component in the allocated memory so the offset for that one is zero. 
+The M component, in comparison, starts at an offset of 4 because we are 
+allocating 4 bytes to the prior component X: bytes 0 through 3 
+specifically. M just occupies one storage element so the next component, 
+B, starts at offset 5. Likewise, component C starts at offset 6. 
```

Comment:
```suggestion
numbered starting at zero. We want the :ada:`X` component to be the very first 
component in the allocated memory so the offset for that one is zero. 
The :ada:`M` component, in comparison, starts at an offset of 4 because we are 
allocating 4 bytes to the prior component :ada:`X`: bytes 0 through 3 
specifically. :ada:`M` just occupies one storage element so the next component, 
:ada:`B`, starts at offset 5. Likewise, component :ada:`C` starts at offset 6. 
```